### ISSUE TITLE
Strategy on adding new options

### ISSUE THREAD

**Issue by czosel** [prettier/plugin-php#120]
I think we should discuss the general procedure on how we handle the wish to add new options.

I started writing this as a comment to #118, but I'd like to discuss the "big picture" in a separate issue instead.

To me, #118 is a nice example to prove @evilebottnawi 's point in the discussion about opening braces (#107, #108):
> I'm not against the option or not specifically this option, the only problem is that adding one option other will want more options, and each will complicate the maintenance of the project and deliver only problems

In this case, I'd advise against introducing a new option, because the number of options we'd have to support if we allow configuration of "this type of" styling decision.
Thinking about it from this standpoint, I'd even support your wish to revert the option we introduced about curly braces and add it back in when we've released a first stable version and the tool has been in use for a while - because then we have an actual community of user that will keep asking for specific options. 
I'd also like to cite two comments by @vjeux which might offer guidance:

> For prettier, I used a combination of my experience writing JavaScript for 10+ years, grepping through the millions of lines of the Facebook codebase to see which style was more popular, reading through various coding styles (eg Airbnb) and what are all the related eslint configs and finally feedback from people complaining that some pattern looked bad.
> My guiding principle was for the code generated by prettier to pass code review as if it was written by a human. I also tried to make all the decisions be as less controversial as possible.

> Also, one thing that really helped is to have someone to make the final decision. This way all the discussions could happen and were useful input but at the end of the day, only the things I wanted would go through.
>I found that there are a handful of useful tactics to manage those discussions:
> *   Ask for specific snippets of code and try to find solutions for those. So many times people argue about very general things but in practice there are concrete fixes you can do for their use case.
> *   Explicitly say that you're not going to make a decision right now on certain issues. One thing that people pick up quickly is to engage in places where the people involve the project are responding. Then once in a while you can sort issues by upvote/comments and figure out what are the most painful things and do a deep dive.
> *   Avoid style trashing. Ideally, you should think deeply about each important style decision you are making and once you do it, actually commit to it. This way you can "close" the issue and whenever this subject comes up in the future say that you have reached a decision and are not looking to spend time thinking about it for now. This way you can focus on the things that still need to be fixed rather than going through the same discussions over an over again wasting everyone's time. This is hard to do in practice but if you can pull it off, it'll help a lot.

(See https://github.com/prettier/plugin-php/issues/59#issuecomment-366744779)

I'm gonna make a suggestion on how i could imagine we handle things, just to keep this issue actionable:

* Until our first release, we don't add any custom options and remain PSR-2 compatible.

After that:

* We only consider a new option if there is a considerable amount of requests to add it
* If necessary, we put the decision up to a vote after a short "cool-down" period

That would also mean that we don't open anything that involves adding options for votes just yet. Of course we can (and should) vote on difficult stylistic decisions that go into the first version.

What do you think?

/cc @mgrip

**Comment by mgrip** [comment]
Thank you for opening - in general I agree. I brought up the brace option specifically because it seems to be very common in a lot of open-source php code I've seen in large, well known projects (not to mention every other coding language 🙄 psr2 why??). It worried me that it might actually push people away from using the plugin. I know for myself, the team of engineers I work with will not buy in to using this as it will mean changing our entire codebase to move opening braces onto new lines.

That being said - I whole-heartedly agree with everything you said in this issue. I would still be in favor of adding the brace option but I totally agree this shouldn't just be my opinion. And going forward for new style issues that come up I agree we should have considerable buy in from the community to prove they're worth adding

**Comment by czosel** [comment]
I could imagine that the brace option will be the first one that we add - but I think it might be a good idea to wait with that until some people have adopted the tool. I'm not too worried about adoption, because:

* we still fill a blank in the PHP tooling landscape
* migration from the (slightly ugly) codebase with PSR2-style linebreaks to the (nicer) style without them can be done completely automatically

Maybe one can also compare it to `--no-semi` for JS: My team at work doesn't like semicolons, but we adopted prettier already before the option existed because the benefits (effortless reproducible formatting) were just outweighing the downsides (semicolons!). A bit later we just migrated to using `--no-semi` when it became available.
Today, i think no one in our team really cares too much about semicolons anymore, because we've all gotten used to the prettier workflow and realized we don't have to add them by hand. Personally, I even prefer to work in a prettier environment without any custom options these days, because then I don't have to keep changing my vim configuration (never really got around to properly integrating `.prettierrc` yet :wink:)

TL;DR: Not supporting some options might have a smaller negative effect on adoption that one would expect, while supporting too many options might have a bigger negative effect in the long run.

**Comment by vjeux** [comment]
For the open brace placement, my feeling is that this is going to be one of those that’s a deal breaker for adopting prettier on PHP if you are on the wrong side of the decision. In JavaScript we got lucky because if you are writing code that way ASI bites you so it’s not common.

For aligning arrays, not only is prettier algorithm not well setup to do that but it is also messes up with diffs and hasn’t been too much an issue in practice for adoption. So I’d say no to it.

**Comment by azz** [comment]
I agree with @czosel's OP, but I can totally see that not supporting the ["one true brace style (1TBS)"](https://en.wikipedia.org/wiki/Indentation_style#Variant:_1TBS_.28OTBS.29) might hinder early adoption. If the majority of projects use 1TBS, maybe it should be the default, with a `--brace-style=k-r` option for PSR compatibility?

**Comment by MichaelDeBoey** [comment]
I think we should indeed start with full PSR-2 compatibility, without options for the first stable release.
Prettier is an **opinionated** code formatter.

I agree that in JS the issue is smaller, but the moment we introduced Prettier in our company, a lot of code changed, because of personal preferences in our codebase.

I'm all in for making it an option after a while (like @czosel said, this could be the very first supported option), but I think we should focus on getting that first stable release out as soon as possible and just focus on the PSR-2 standard since it's the most widely used standard in the PHP community.

**Comment by czosel** [comment]
Just went through the first two pages of most starred PHP projects on GitHub:
### 1TBS
Wordpress
Phabricator

### K-R
Laravel
Symfony
CodeIgniter
Faker
composer
PHPMailer
Guzzle
Yii
Matomo
cphalcon
PHPUnit
grav
Carbon
Cachet
mobiledetect
CakePHP

I think we'd be ok with going for the PSR-2 compatible style in our first release.

**Comment by MichaelDeBoey** [comment]
Like I said: start with PSR-2 as default, go to PSR-12 by default when it's accepted

Eventually we can add options to have other standards or add extra options like for the opening bracket...

**Comment by cmancone** [comment]
The thing I like about this (other than the fact that it automates code styling, of course), is the fact that it *is* opinionated.  For me, that's a big plus.  Anytime there is a question of "why do we format it like this?" the answer is "Because that's what our tooling does, and we use our tooling so we don't have to think about it".  It takes the focus (of the coders) away from the minutiae and arguments over where to put braces and towards the benefits of "I don't even have to think about it anymore".  Providing options inherently works against that goal.

That being said, I can see the need for a few simple options to ease the transition for people who have their own strong opinions.  I can also see the counter argument though - if someone has strong opinions about how their code should be formatted, then they probably just don't want an opinionated formatter, and that's okay too.  I might propose a simple suggestion (and I won't be insulted if this isn't how you guys choose to do things):

 1. Don't worry about having any options in the initial release phases
 2. Identify a small number of high-controversy options to support for user convenience (probably Brace location, maximum line length, align arrows/don't align arrows, and number of spaces - but that list may be too long)
 3. Bake those in as the *only* allowed options
 4. Automatically refuse all future requests for options.

This seems to largely be in line with what others are saying here.  Again, from my perspective I don't actually care what the format is.  We use inline opening braces in our code style (I find having it on the next line leaves too much empty space), but even if you only ever supported PSR-2 and its silly next-line-opening-brace (that's right, I'm *trying* to insult people here :p ), I would still be **extremely** happy to use this.

**Comment by SidIcarus** [comment]
I, like those who's commented above, hold to the opinion of code format consistency is most necessary and numerous options can bring on unnecessary [bikeshedding](https://en.wiktionary.org/wiki/bikeshedding). 
  
With that, I've swayed my team to adhere to [JS-Standard-Style](https://standardjs.com/) - with semi colons (because our php-js interpolation is too high), and the html / js-php plugins. Their ruleset is established via eslint. It'd be nice to have pure PHP follow suite.
Enter prettier or phpcs/phpcbf; emulating JSSS via the latter has been more hassle-some then not. 
### Proposal
> Use eslint-plugin with this plugin to establish ruleset from JSSS. Apply non-js specific
rules to php files.

This would alleviate having to think about cross-language styling.

Thoughts?

**Comment by aboyton** [comment]
I completely agree with hating options. I read everyone asking for options in Prettier for JavaScript and cringe. "Just one more option" very very quickly leads to way too many options and goes against the whole point of Prettier meaning "I don't have to care about formatting/style".

That said, I'm also one of those people who really really prefers 1TBS. Running prettier over our codebase at work shows a bunch of small formatting differences between our style guide and what Prettier does but I suspect our company could get over them all. The one that's a real blocker for us is 1TBS. We're generally following Phabricator's coding guidelines and like pretty well everyone else in this thread we really can't understand why PSR does what it does.

Now that we've released an alpha version, thoughts on adding support for 1TBS? Having this almost certainly a blocker for us adopting Prettier for PHP.

**Comment by cmancone** [comment]
Can you clarify your comment @aboyton ?  From my understanding 1TBS can mean a few different things.  Here is a relevant example:
Input:
```php
<?php
class test {
    public function another_test() {
        if (true)
            print 'hi';
    }
}
```
Output:
```php
<?php
class test
{
    public function another_test()
    {
        if (true) {
            print 'hi';
        }
    }
}
```
Relevant notes:
 1. Opening brace for classes are on a new line
 2. Opening brace for functions are on a new line
 3. Opening brace for other blocks are on the same line
 4. Braces are auto-inserted for one-line statements

I'm guessing that you support items 3&4 but disagree with 1&2.  We use same-line braces in all cases and that would be my preference (and I would understand brace-position being an option in the future), but the current setup wouldn't stop me from adopting prettier for PHP even if there was never an option otherwise.

**Comment by aboyton** [comment]
Ah definitely @cmancone. I was referring to https://github.com/prettier/plugin-php/issues/120#issuecomment-368318324 in my naming (although that was far from clear).

I was referring to 1 and 2 so that we'd instead put the braces on the same line. I'm very happy with 3 and 4 how they are.

As for if not having 1 and 2 and options holding us back from adopting Prettier, it's hard to actually say if they actually would stop us, or even if having them would mean we would adopt prettier for our PHP code at work. It's impossible to predict the future with such things and I think a lot of people were against Prettier in various communities before they actually tried it and realised how nice it was to no longer think about formatting code and just not caring about if prettier matched their exact personal preferences.

This comment sums up my experience with prettier wonderfully. https://github.com/prettier/prettier/issues/3437#issuecomment-360219895
> FWIW I've long stopped caring about things like this and I can't recommend that enough

That said, we're pretty heavily invested in our current PHP linting tools at work and the idea of having JavaScript formatting PHP rather than PHP formatting PHP doesn't sit well with those in our team that deal only with PHP. So the question comes can I sell Prettier as a better solution to code formatting than our current tools and I think that is going to be a much harder sell while 1 and 2 are how they are.

**Comment by SidIcarus** [comment]
@aboyton by could you clarify what you mean by 
> the idea of having JavaScript formatting PHP rather than PHP formatting PHP

My two interpretations are:
1. Having the tool that formats PHP be written in JS rather then PHP
2. Having a JS `format` for PHP rather then a PHP `format` for PHP 

My thoughts on both:
1. If the barrier to access and usage is low (easily integrated into workflow & environment / easy to use), why is what it is implemented in a concern?
2. Code is code. A common issue all styles try to address is a lack of consistency. Even though a member of a team may not develop in multiple languages, if that is a scenario for other developers in that persons team, having a consistency that will allow everyone to more easily view and comprehend code is nothing but beneficial. 
2.1 a common example:  consistency between HTML, CSS, & JS.

**Comment by evilebottnawi** [comment]
@SidIcarus can't understand, you want what `prettier-php` was written on `php`?

**Comment by aboyton** [comment]
As risk of derailing the conversation, my point was simply that generally speaking tools for a language are commonly written in that language itself. It helps both for people working on the project, but also for people using the tools. If we integrate Prettier into our PHP workflow at work then for them to be able to submit code they need to have Node installed. I doubt this will be a problem for me personally as our repositories contain both PHP and JavaScript together but it will annoy others.

As a case in point, Haskell apparently use Phabricator for submitting changes (like we do at work) and so to contribute to Haskell you need to have PHP installed. Not everyone seems happy with that—https://ghc.haskell.org/trac/ghc/wiki/WhyNotPhabricator Similarly the fact that Flow is written in OCaml appears to have hindered it's adoption somewhat (where as TypeScript is written in TypeScript).

I'm not meaning to derail the conversation or say that this project shouldn't exist, on the contrary I'm really hoping we can take everything we've loved about using Prettier for our JavaScript and TypeScript and get it for our PHP code, I'm just saying that not everyone I work with will be impressed needing Node to format PHP code (although I'm hoping the benefits outweigh the extra dependancy and hurdle of contributing to the project).

**Comment by cmancone** [comment]
@aboyton I think you're talking about a few different issues at the same time: the fact that this is written in javascript, and the fact that the formatting for braces doesn't currently match the coding standards for your team.

Regarding the use of javascript, obviously there is nothing to be done about that fact now.  For me that isn't even a disadvantage: I found this because I was also looking for a tool to format our typescript code, so the fact that I could use the same tool to format both the same way is a big win for me.  If the fact that this is written in javascript makes it harder to sell to your team, then that is unfortunate, but ultimately if this isn't a fit for your team then it isn't a fit for your team.  Personally I think these guys made a great decision to build off of the prettier system that was already in place.  I'm a big fan of using the right tool for the right job, and I really couldn't imagine trying to write something like this for PHP.  That might be a bit of my personality though.  I'm a big fan of PHP for the web, but my love of it ends there - I've always used python or other languages for tooling because I find them to be much better fits, and the extra dependencies have never bothered me.

In the meantime, every decision made by the contributors here are going to appeal to some people while driving away others.  Because braces are probably on the top of the list of things that people take very seriously, I would vote for braces being on the top of the list of things to provide options for, presuming that the contributors decide that they want to provide some options.  I actually agree with your preference: I'd rather have all opening braces on the same line.  For me though the benefits of using prettier far outweigh that minor detail of formatting, so I'll be using this as soon as it is ready, and if they decide to add an option for braces down the line then I'll just be that much happier.

**Comment by evilebottnawi** [comment]
In theory (very very very big theory) we have `babel`, and we have `php-parser` on `nodejs`, we can write something like this https://babeljs.io/php/ but in other direction. Prettier already have js bundle what works in browser without node env and in theory can be compiler to PHP. I know that this is not an easy task, but if someone create babel-plugin/babel-preset for this we can create bundle for php users without js. Personally for me it's a bit strange and I will not do this, but I'm open to such an opportunity, everything is in your hands.

**Comment by aboyton** [comment]
@cmancone I'm definitely talking about two things, and I didn't want to derail the conversation about options by taking about the fact the formatting PHP will require node as a dependancy.

I think I agree with you completely, I think building this on top of Prettier far outweighs the cons (both the extra dependancy but also making people who want to contribute need to know JS as well as PHP). I've loved what Prettier has brought to my coding (not thinking about style at all is incredibly liberating) and am looking forward to the same happening for PHP.

What I'm more concerned about is the option to put braces on the same line. I think most of us agree on this being nicer and would prefer it that way and are torn between not adding too many options (ideally zero) and no understanding why PHP's style guidelines are the way that they are 😄

**Comment by evilebottnawi** [comment]
@aboyton can you explain example of code?

**Comment by aboyton** [comment]
See https://github.com/prettier/plugin-php/issues/120#issuecomment-380264918

Given
```php
<?php
class test {
    public function another_test() {
        if (true)
            print 'hi';
    }
}
```
I'd love to have output:
```php
<?php
class test {
    public function another_test() {
        if (true) {
            print 'hi';
        }
    }
}
```

**Comment by evilebottnawi** [comment]
@aboyton I think it is not problem.
1. Create issue with request option for this (add example as above) and let's wait people will vote.
2. You can gather information of popular code styles and popular libraries/frameworks around this problem, if a lot of people use this code style this will be a weighty argument.
3. If we reject this option you can implement rule for `phpcs` to fix it (i.e. run `prettier` and when run `phpcs`, `prettier` use same approach). Also improve documentation so that future users can easily implement sniffs.

**Comment by aboyton** [comment]
I think the main question here isn't "should we add this option in particular" but "should we add this option now or should we wait (or never add any options)". If we're adding any options I think pretty well everyone on this thread agrees that it would be this option, see for example:

@mgrip https://github.com/prettier/plugin-php/issues/120#issuecomment-368042456
@czosel https://github.com/prettier/plugin-php/issues/120#issuecomment-368048042
@vjeux https://github.com/prettier/plugin-php/issues/120#issuecomment-368058086
@MichaelDeBoey https://github.com/prettier/plugin-php/issues/120#issuecomment-368235815

So I can open a new issue if you would like, but I don't think that will help anything as the question is "do we want to start adding options now" which is what this issue is about.

**Comment by cmancone** [comment]
@aboyton Actually I just realized that there is pretty much already an open issue regarding brace location specifically: #107 and #108, as referenced in the original issue here.

**Comment by czosel** [comment]
I agree, let's close this!
